### PR TITLE
add legend ordering for new scenarios in `trajectory_alignment.js`

### LINKED ISSUE #52: bug: volume trajectory plot seems to not render `2022Q4` scenarios correctly
The development version of the interactive report, using the `2022Q4` and `2023Q4` timestamps, seems to not render the volume trajectory plot as expected. 

I am uncertain if this is an issue with this particular repo `pacta.portfolio.report` or one of the workflows, but I will flag it here, and we can transfer it or close it as we determine what is causing it.

FYI @cjyetman @AlexAxthelm 

### 2022Q4:
Development report that exhibits the issue:
https://pactadatadev.blob.core.windows.net/ghactions-workflow-transition-monitor-results-reports/rmi_pacta_2022q4_general-20240321T113136Z/EN/other/1/working_dir/50_Outputs/rmi_pacta_2022q4_general/report/index.html

Screenshot:
Screenshot 2024-03-21 at 12 49 09

### 2023Q4:
Development report that exhibits the issue:
https://pactadatadev.blob.core.windows.net/ghactions-workflow-transition-monitor-results-reports/rmi_pacta_2023q4_general-20240321T113124Z/EN/other/1/working_dir/50_Outputs/rmi_pacta_2023q4_general/report/index.html

Screenshot:
Screenshot 2024-03-21 at 12 50 41

### LINKED ISSUE #53: need to add legend ordering for new scenarios in 2022Q4 and 2023Q4 data
2022Q4 and 2023Q4 datasets include new scenarios that have not previously been included

https://github.com/RMI-PACTA/workflow.data.preparation/blob/680eda103914219b756e9630315378ae83dbd4f8/config.yml#L131
``` yaml
  scenario_raw_data_to_include: ["geco_2022", "isf_2021", "weo_2022"]
```

https://github.com/RMI-PACTA/workflow.data.preparation/blob/680eda103914219b756e9630315378ae83dbd4f8/config.yml#L171
``` yaml
scenario_raw_data_to_include: ["geco_2023", "isf_2023", "weo_2023"]
```

``` r
# 2022Q4
unique(pacta.scenario.preparation::geco_2022$scenario)
#> [1] "1.5C"      "NDC_LTS"   "Reference"
unique(pacta.scenario.preparation::isf_2021$scenario)
#> [1] "NZE"
unique(pacta.scenario.preparation::weo_2022$scenario)
#> [1] "APS"      "NZE_2050" "STEPS"

# 2023Q4
unique(pacta.scenario.preparation::geco_2023$scenario)
#> [1] "1.5C"      "NDC-LTS"   "Reference"
unique(pacta.scenario.preparation::isf_2023$scenario)
#> [1] "1.5°C"
unique(pacta.scenario.preparation::weo_2023$scenario)
#> [1] "STEPS"    "APS"      "NZE_2050"
```

therefore we will need to include these new scenarios in the legend ordering for the trajectory volume plot...
https://github.com/RMI-PACTA/pacta.portfolio.report/blob/24ebb45079153fb62d6080572830b02afb0fdfc2/inst/js/trajectory_alignment.js#L381-L399

ℹ️ FYI @MonikaFu @AlexAxthelm @jdhoffa

### PR DESCRIPTION
**Pull request by cjyetman**
- closes #53
- closes #52 (hopefully)

- [x] needs @MonikaFu's review for the colors and functionality
- [x] needs @jdhoffa's review for the scenarios and proper ordering

### CODE REVIEW

**Review by jdhoffa — APPROVED** [review]
💜

File: inst/js/trajectory_alignment.js [hunk 1]
```diff
@@ -390,6 +390,42 @@ class trajectory_alignment {
...
+          .domain(["production", "1.5\xc2\xb0C", "worse"])
+          .range(["black", "#9cab7c", "#e07b73"])
+          legend_order = ["worse", "1.5\xc2\xb0C"];
+        ;
+      } else if (selected_source == "WEO2023") {
+        color = d3.scaleOrdinal()
+          .domain(["production", "NZE_2050", "APS", "STEPS", "worse"])
+          .range(["black", "#9cab7c", "#FFFFCC", "#fde291", "#e07b73"])
+          legend_order = ["worse", "STEPS", "APS", "NZE_2050"];
+        ;
```
Comment (jdhoffa) [inline inst/js/trajectory_alignment.js:428]: The WEO 2023 scenario seems to be called `NZE` here (in contrast to WEO 2022 where it is`NZE_2050`):
``` r
library(pacta.scenario.preparation)

weo_2023 |> 
  dplyr::distinct(scenario)
#> # A tibble: 3 × 1
#>   scenario
#>   <chr>   
#> 1 STEPS   
#> 2 APS     
#> 3 NZE

weo_2022 |> 
  dplyr::distinct(scenario)
#> # A tibble: 3 × 1
#>   scenario
#>   <chr>   
#> 1 APS     
#> 2 NZE_2050
#> 3 STEPS
```

<sup>Created on 2024-03-25 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Review by jdhoffa — CHANGES_REQUESTED** [review]
See comment

File: inst/js/trajectory_alignment.js (hunk 1 above)
Comment (cjyetman) [reply inst/js/trajectory_alignment.js:428]: I thought that `WEO2023` would have `NZE_2050` after https://github.com/RMI-PACTA/pacta.scenario.preparation/pull/140 no?

Comment (jdhoffa) [reply inst/js/trajectory_alignment.js:428]: Oh bizarre, maybe I'm using an old version installed locally

Comment (cjyetman) [reply inst/js/trajectory_alignment.js:428]: related https://github.com/RMI-PACTA/workflow.transition.monitor/pull/288

**Review by jdhoffa — DISMISSED** [review]
Retracting my comment! I was using an outdated version of `pacta.scenario.preparation`. 

This LGTM

**Comment by MonikaFu** [comment]
@cjyetman thanks! I will give it a look tomorrow morning!

**Comment by MonikaFu** [comment]
@cjyetman I run the code and looked into the report generated and the problem seems still not to be resolved. I looked into the .js just to make sure that I am using the code from this PR and I am using the same version but the plots don't render correctly. For example, for ISF there are no scenarios

image

And for WEO2023 we see a black area while there should only be colourful scenario areas.

image

**Comment by jdhoffa** [comment]
> @cjyetman I run the code and looked into the report generated and the problem seems still not to be resolved. I looked into the .js just to make sure that I am using the code from this PR and I am using the same version but the plots don't render correctly. For example, for ISF there are no scenarios
> 
> image
> And for WEO2023 we see a black area while there should only be colourful scenario areas.
> 
> image

@cjyetman just jumping in here, I could be wrong, but I believe this is because we need to re-run data-prep to pick up the changes in scenario name in `pacta.scenario.preparation`? At least for the case of WEO?

ISF I'm not sure what the issue is

**Comment by cjyetman** [comment]
I did advise @MonikaFu that the WEO NZE scenario needs to be temporarily, manually adjusted to `WEO_NZE` instead of `WEO_NZE_2050` in the workflow.transition.monitor project configs to work with the current data.prep outputs that we have. That appears to be the problem in these plots because `NZE` is missing. I also don't know if the explicitly defined "worse" is actually necessary, I just tried to follow the previous pattern. My expectation was that "worse" defined the color of the area beyond the worse scenario, but I don't really know how this works.

With ISF2023, I also don't know if it really needs "worse". Also no idea why it's not picking up "1.5°C".

If this doesn't work, we can close this ticket and I guess @MonikaFu will have to figure it out.

**Comment by MonikaFu** [comment]
I have adjusted the parameter as advised but still see this issue. If I look into the parameters used in the code I see `select_scenario = "WEO2023_NZE"` which is what we want with current data.prep outputs right?

You always need "worse" as far as I remember in the way the colour assignment works in this plot. But would need to dive deeper to recall.

Yes, I guess since it seems this PR does not manage to solve the issues with adding new scenarios I probably need to take over.

**Comment by jdhoffa** [comment]
Then is the intention to close this PR?

**Comment by MonikaFu** [comment]
> Then is the intention to close this PR?

I'm working within this PR as the changes @cjyetman made are needed but there is more that needs to change for the scenarios to be picked up.

**Comment by jdhoffa** [comment]
@MonikaFu is this ready for review? or was the review request a mistake?

**Comment by MonikaFu** [comment]
> @MonikaFu is this ready for review? or was the review request a mistake?

it was kind of a mistake. I did not want it to have the 'approved' status so I re-requested a review but I only realised after that it will ask you for a review of course. It is not ready for review that is why I put it into 'draft' mode.

**Comment by MonikaFu** [comment]
@jdhoffa I believe this is solved now. The 'WEO2023' source still does not work properly because the plot expects a scenario called 'NZE_2050' but in the current version of the data it is 'NZE'. From what I understand this is going to be fixed and the proper name should be 'NZE_2050' so I am not going to change it.

**Comment by MonikaFu** [comment]
@cjyetman maybe you would like to have a look at it as well?

**Review by MonikaFu — APPROVED** [review]
LGTM but some of the changes were made by me so I believe someone else should have another look.

**Comment by cjyetman** [comment]
Seems to function as expected, minus the NZE_2050 situation. But... is "worse" supposed to appear black? and is there something wrong with the order of "worse"?
Screenshot 2024-04-05 at 13 44 08
Screenshot 2024-04-05 at 13 44 21

also, not sure if this is related, but I get 2 JavaScript errors that prevent the trajectory plots from displaying until I change one of the selectors...
Screenshot 2024-04-05 at 13 46 11
Screenshot 2024-04-05 at 13 46 25

**Comment by jdhoffa** [comment]
> Seems to function as expected, minus the NZE_2050 situation. But... is "worse" supposed to appear black? and is there something wrong with the order of "worse"? Screenshot 2024-04-05 at 13 44 08 Screenshot 2024-04-05 at 13 44 21
> 
> also, not sure if this is related, but I get 2 JavaScript errors that prevent the trajectory plots from displaying until I change one of the selectors... Screenshot 2024-04-05 at 13 46 11 Screenshot 2024-04-05 at 13 46 25

Is this an artifact of the `NZE_2050` issue (and a mismatch between expected and actual scenarios?)

Does "worse" still appear black if you try a different scenario?

**Comment by MonikaFu** [comment]
Worse should not appear at all but it appears because of the mismatch between expected and actual scenario names. No idea about the

**Comment by MonikaFu** [comment]
@cjyetman makes sense. Then I will leave it open and marked as 'blocked' on ADO until we get a new version of `data.prep`.

**Comment by jdhoffa** [comment]
Depends on:
https://dev.azure.com/RMI-PACTA/2DegreesInvesting/_workitems/edit/10505

**Comment by cjyetman** [comment]
For what it's worth, I jumped through all of the hoops (I think), and with @MonikaFu's additions, this appears to work as expected! 👍🏻

**Comment by MonikaFu** [comment]
@cjyetman does it mean that you tested with the right data? And it works? Can we merge?

**Comment by cjyetman** [comment]
> @cjyetman does it mean that you tested with the right data? And it works? Can we merge?

Yeah, I manufactured an inputs dataset that has the expected scenario name. There were still a few problems (like the JS errors mentioned above), but I believe those are separate issues. I think this PR is safe, but I leave it up to you to decide if/when to merge.

**Pull request merged**